File: 강의 자료/02-알고리즘/01~02-완전 탐색/문제별 코드/1759-암호 만들기/solution.py
import sys
input = sys.stdin.readline
m, n = map(int, input().split(' '))
chars = sorted(input().strip().split(' '))
used = [0] * n
selected = [0] * m

def is_vowel(x: str):
    return x in "aeiou"

def rec_func(k):
    if k == m:
        vowel, consonant = 0, 0
        for x in selected:
            if is_vowel(chars[x]):
                vowel += 1
            else:
                consonant += 1

        if (vowel >= 1) and (consonant >= 2):
            for x in selected:
                print(chars[x], end='')
            print()

    else:
        st = -1 if k == 0 else selected[k - 1]
        for i in range(st + 1, n):
            selected[k] = i
            rec_func(k + 1)
            selected[k] = 0

rec_func(0)

File: 강의 자료/02-알고리즘/01~02-완전 탐색/문제별 코드/1759-암호 만들기/solution.cpp
#include <iostream>
#include <algorithm>
typedef long long int ll;
using namespace std;

int N, M, selected[16];
char chars[16];
void input() {
	cin >> M >> N;
	for (int i = 1; i <= N; i++) cin >> chars[i];
}

bool isVowel(char x) {
	return x == 'a' || x == 'e' || x == 'i' || x == 'o' || x == 'u';
}

void rec_func(int k) {
	if (k == M + 1) {
		int vowel = 0, consonant = 0;
		for (int i = 1; i <= M; i++) {
			if (isVowel(chars[selected[i]])) vowel++;
			else consonant++;
		}
		if (vowel >= 1 && consonant >= 2) {
			for (int i = 1; i <= M; i++) cout << chars[selected[i]];
			cout << "\n";
		}
		return;
	}
	for (int cand = selected[k - 1] + 1; cand <= N; cand++) {
		selected[k] = cand;
		rec_func(k + 1);
		selected[k] = 0;
	}
}
int main() {
	ios::sync_with_stdio(false); cin.tie(0); cout.tie(0);
	input();
	sort(chars + 1, chars + 1 + N);
	rec_func(1);
	return 0;
}